File: Chapter- 14/topological_sorting.cpp
#include<iostream>
#include<stack>
#define N 6

using namespace std;

int graph[N][N] = {
   {0, 0, 0, 0, 0, 0},
   {0, 0, 0, 0, 0, 0},
   {0, 0, 0, 1, 0, 0},
   {0, 1, 0, 0, 0, 0},
   {1, 1, 0, 0, 0, 0},
   {1, 0, 1, 0, 0, 0}
};

void Topological(int u, bool visit[], stack<int>&sk) {
   visit[u] = true;

   for(int v = 0; v<N; v++) {
      if(graph[u][v]) {
         if(!visit[v])
            Topological(v, visit, sk);
      }
   }
   sk.push(u);
}

void topological_sort() {
   stack<int> sk;
   bool vis[N];

   for(int i = 0; i<N; i++)
      vis[i] = false;

   for(int i = 0; i<N; i++)
      if(!vis[i])
         Topological(i, vis, sk);

   while(!sk.empty()) {
      cout << sk.top() << " ";
      sk.pop();
   }
}

int main() {
   cout << " Topological sort is : ";
   topological_sort();
   return 0;
}

File: Chapter- 14/topological_sorting.py
from collections import defaultdict

class Graph:
    def __init__(self, vert):
        self.graph = defaultdict(list)
        self.V = vert

    def add_Edge(self, u, v):
        self.graph[u].append(v)

    def Topological(self, v, visit, stack):

        visit[v] = True

        for i in self.graph[v]:
            if visit[i] == False:
                self.Topological(i, visit, stack)

        stack.append(v)

    def topological_Sort(self):
        visit = [False] * self.V
        stack = []

        for i in range(self.V):
            if visit[i] == False:
                self.Topological(i, visit, stack)
        print(stack[::-1])


gr = Graph(4)
gr.add_Edge(3, 2)
gr.add_Edge(3, 0)
gr.add_Edge(2, 0)
gr.add_Edge(2, 1)
gr.add_Edge(0, 3)
gr.add_Edge(1, 1)

print("Topological Sort is : ")

gr.topological_Sort()